### PR TITLE
crypto: add a couple of tests for ed25519 signatures before/after verification

### PR DESCRIPTION
**Pull request by cce**
## Summary

Now that the consensus upgrade for #3031 is complete this pushes a couple of unit tests related to signatures that were not valid before the batch verification upgrade, but are valid now.

Related: #3781

## Test Plan

Adds two new tests.

### CODE REVIEW

**Comment by algoidan** [comment]
From my understanding,  the purpose of those tests is to check `ed25519verify` TEAL opcode under specially crafted public keys, right?
if so, what is the difference between those two tests? (I do understand that they have different logic, but the test tries to fail the opcode)

Is there a way to break one test without the other?

**Review by algoidan — COMMENTED** [review]
From my understanding, the purpose of those tests is to check ed25519verify TEAL opcode under specially crafted public keys, right?
if so, what is the difference between those two tests? (I do understand that they have different logic, but this is not the focus of the test)

Is there a way to break one test without the other?

File: data/transactions/logic/evalCrypto_test.go [hunk 1]
```diff
@@ -173,6 +175,154 @@ ed25519verify_bare`, pkStr), v)
...
+				{1000000, false, false}, // EnableBatchVerification = false, can't withdraw 1 algo
+				{2000000, false, true},  // EnableBatchVerification = false, withdraw 2 algos
+				{1000000, true, true},   // EnableBatchVerification = true,  withdraw 1 algo
+				{2000000, true, false},  // EnableBatchVerification = true,  can't withdraw 2 algos
+			} {
+				t.Run(fmt.Sprintf("%+v", tc), func(t *testing.T) {
+					var txn transactions.SignedTxn
+					txn.Lsig.Logic = ops.Program
+					txn.Lsig.Args = [][]byte{data[:], vec[2], vec[1]}
+					sb := strings.Builder{}
```
Comment (jannotti) [inline data/transactions/logic/evalCrypto_test.go:312]: I don't see where you're putting `sb` in to trace.  But you shouldn't have to anyway.  `defaultEvalParams` inserts a `string.Builder` for you, so just change to evalParams.Trace.

File: data/transactions/logic/evalCrypto_test.go [hunk 2]
```diff
@@ -173,6 +175,154 @@ ed25519verify_bare`, pkStr), v)
...
+		t.Run(fmt.Sprintf("valid,vec=%v", vecs[i]), func(t *testing.T) {
+			// logic sig that lets you withdraw different amounts if EnableBatchVerification=true
+			/* (||
+			    (&& (ed25519verify (arg 0) (arg 1) (arg 2))
+			      (== (txn Amount) 1000000))
+			    (&& (! (ed25519verify (arg 0) (arg 1) (arg 2)))
+			      (== (txn Amount) 2000000))
+			   )
+			*/
+			ops, err := AssembleStringWithVersion(`arg 0
```
Comment (jannotti) [inline data/transactions/logic/evalCrypto_test.go:273]: I like to use `testProg` for assembling in tests. It checks the error for you and does not return it, it disassembles and reassembles to check equality giving us more tests for free, and does some logging if things don't work out.

**Review by jannotti — APPROVED** [review]
Seems good, willing to merge as is, let me know if you want to make any of the changes I mentioned. (the sb trace is probably the most worthwhile)

**Comment by cce** [comment]
>From my understanding, the purpose of those tests is to check ed25519verify TEAL opcode under specially crafted public keys, right?
>if so, what is the difference between those two tests? (I do understand that they have different logic, but this is not the focus of the test)
>Is there a way to break one test without the other?

There is no difference between the tests from the POV of the ed25519verify opcode operation, in the second one I was working on a transaction that would result in an account balance change, but it is not necessary if we want to just assert the new ed25519 behavior.

**Comment by cce** [comment]
Unnecessary after #3781 merged

**Pull request closed without merging**